notes/conds.md

```diff
@@ -0,0 +1,45 @@
+# delegation clause
+
+In 0.x verison of UCANs has an open-ended `nb` field that most users use to constraints delegated capabilities. Open ended nature makes interop very difficult as how to interpert those constraints are not covered by spec. In 1.0 we would like to define standard but extensible constarints language so that implementations can interop unless user space extensions are utilized. By allowing user space extensions of the constraint language we also allow user space to innovate at the cost of interop.
+
+## s-expression based language
+
+### Operators
+
+| Operator     | Function Signature           | Description                             | Example                            |
+| ------------ | ---------------------------- | --------------------------------------- | ---------------------------------- |
+| `["a", "b"]` | `[String] -> Path`           | Map key (or array index) selector chain | `["key", 2]`                       |
+| `==`         | `Expr -> Expr -> Bool`       | Equality                                | `["==", ["key"], 1]`               |
+| `<`          | `Path -> Num -> Bool`        | Less than                               | `["<", ["key"], 2.3]`              |
+| `<=`         | `Path -> Num -> Bool`        | Less than or equal                      | `["<=", ["key"], 2.3]`             |
+| `>`          | `Path -> Num -> Bool`        | Greater than                            | `["<", ["key"], 2.3]`              |
+| `>=`         | `Path -> Num -> Bool`        | Greater than or equal                   | `["<=", ["key"], 2.3]`             |
+| `not`        | `Expr -> Bool`               | Negation                                | `["not", ["<=", ["key"], 2.3]`     |
+| `some`        | `Collection -> Expr -> Bool` | Collection predicate                    | `["some", ["==", ["key"], 42]]`     |
+| `every`        | `Collection -> Expr -> Bool` | Collection predicate                    | `["every", ["==", ["key"], 42]]`     |
+| `match`    | `Path -> Pattern -> Bool`    | String wildcard                         | `["match", ["key"], "*@foo.com"` |
```

Comment:
```suggestion
| `match`    | `Expr -> Expr -> Bool`    | String wildcard                         | `["match", ["@", "/to"], "*@foo.com"]` |
```

Reply:
An alternate version using strings:

```js
["match", ["path", "to.[+]..bar.[*].email"], "*@fission.codes"]
```

This selector puts the some/every inline, and uses characters familiar from regexes. The one above says:

`"match" "to.{one or more children}.{recursive descent}.{all children}.email" "ends with @fission.codes"`

Reply:
I've also played with some kanren/core.logic inspired `fresh` bindings, but I genuinely think that we can punt on that for now. However, if we were to have them, they can also be made to work nicely with the selector synatx:

```js
["fresh", ["X", "Y"],
  ["match", ["path", "to.[+]..bar.[X].email"], "Y"],
  [">", "X", "Y"]
]
```

Which is kind of a ridiculous example, but basically lets you unify on both the value and parts of the selector.

Reply:
Also taking a page out of SQL's book, we could do something like this to make it really unambiguous (though it's more verbose in simple cases)

```js
{
  iss: "did:key:alice",
  // ...
  "select": {
    "X": "args.to.[+]..bar.[X].email",
    "Y": "args.from"
  },
  "in": [
    ["==", X, 42],
    [">=", X, Y]
  ]
}
```

Reply:
Expressing the same in a karnen-like format:

```js
{
  iss: "did:key:alice",
  // ...
  cond: [
    ["fresh", ["?email", "?from"],
                      // Selector
                      // vvvvvvvvvvvvvvvvvvvvvvv
       ["find", ["args", "to.[+]..bar.[*].email", "?email"]],
       ["find", ["args", "from", "?from"]],
                      /  ^^^^^^
                      // Also a selector
       ["==", "X", 42],
       [">=", "X", "Y"]    
  ]
}
```

Reply:
☝️ I actually don't hate this. Really simple cases can still be minimal:

```js
["args", "to.[+]..bar.[*].email", 42]
```

but anything more complex uses logic variables

Reply:
It also eliminates the need for `==` in a lot of cases, because we can use pattern matching in the really basic cases

Reply:
I personally prefer datomics implicit bindings. That way you can't undefined references. I would also prefer if we don't overload strings too much. Specifically I have started using `[]` to reference context (which generally assumes `args` of the invocation) and `["/foo"]` or `["/foo/0/bar"]` to select using JSON path. I think doing similar thing with variables would fit really well e.g `["?foo"]`

Reply:
Specifically this is ambiguous if we just use strings

```js
["==", "/foo",  "/bar"],
```

Dose that mean `args.foo == args.bar` or does it mean `"/foo" === "/bar"` or `orgs.foo === "/bar"`